### DISCUSSION TITLE
Macro

### DISCUSSION THREAD

**Discussion by MohammadMD1383** [MohammadMD1383/Actionable#91] [Announcements]

**Comment by MohammadMD1383** [comment]
# First Beta Is Here!

![image](https://user-images.githubusercontent.com/69088224/218274752-5918e088-98b4-4112-b44c-ae0d650b2072.png)

1- You can open macro settings from `Settings > Tools > Actionable > Macro` or simply from the shortcut as shown above

2- Your macros will be listed there and you can assign shortcuts to them, like past.

![image](https://user-images.githubusercontent.com/69088224/218274921-d20cf7fe-f7b5-4cc7-a7b7-6cfd87add09f.png)

1- You can see where your macros are stored. (each macro is stored in a `.txt` file in the specified folder)

2- You can see all of your macros

3- You can edit your macro file here
  * all changes are immediately written to the file
  * `cancel` or `reset` buttons cannot get you back to the previous state, even `apply` and `ok` does nothing except that they refresh the macros (if you choose `cancel` macros will be updated on next ide startup)

# Whats new?

The new macro engine is supposed to have these benefits:
* easier macro create/delete/edit
* better macro editor experience
* template engine like *live templates* (for now `$SELECTION$` and `$0$`(final caret position) is available)
* better and more useful code generation

# Example
before `alt+c` (custom shortcut assigned to a macro):
![image](https://user-images.githubusercontent.com/69088224/218275823-90fa01bb-c697-47cd-8601-36085e8b176a.png)

after:
![image](https://user-images.githubusercontent.com/69088224/218275819-96ba3269-7699-4a9e-b43b-f515c245c350.png)

# Final Speech
* creating/deleting macro `.txt` files are not yet supported and you have to manually create/delete them
* __Don't forget to test this new feature and give feedback! Just visit releases page for latest nightly build__
* After upgrading to latest plugin version, older macros will be discarded, so take a copy of them if you still need them

**Comment by j-a-s-o-n-g** [comment]
Will this have sublime text macro capabilities where it can write text based on regex expressions or # of characters written? Example:

```

|

```

```
-
a|
-
```

```
--
ad|
--
```

```
<snippet>
  <content><![CDATA[
${1/./-/g}
${1:}
${1/./-/g}
]]></content>
  <tabTrigger>aaa</tabTrigger>
</snippet>
```

**Comment by MohammadMD1383** [comment]
this looks to be a complex premium feature!

the goal of actionable macro is to insert a text with a shortcut so this is not covered in this issue, but this can be a separate feature request and is welcomed!

**Comment by j-a-s-o-n-g** [comment]
Ok, I made it a new feature request issue, but definitely not something urgent.

**Comment by MohammadMD1383** [comment]
## Changelog

* added add/remove buttons to settings page
* added word/element support
* handle replacement correctly

> This feature is going to be unveiled in v3.17.0, but this thread won't be closed and discussion about this feature is still active to improve it!
